spring-kafka/src/test/java/org/springframework/kafka/listener/ContainerEnforceRebalanceTests.java

```diff
@@ -0,0 +1,144 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.listener;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+
+import java.time.Duration;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.apache.kafka.common.TopicPartition;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.kafka.annotation.EnableKafka;
+import org.springframework.kafka.annotation.KafkaListener;
+import org.springframework.kafka.config.ConcurrentKafkaListenerContainerFactory;
+import org.springframework.kafka.config.KafkaListenerEndpointRegistry;
+import org.springframework.kafka.core.ConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaConsumerFactory;
+import org.springframework.kafka.core.DefaultKafkaProducerFactory;
+import org.springframework.kafka.core.KafkaTemplate;
+import org.springframework.kafka.core.ProducerFactory;
+import org.springframework.kafka.test.EmbeddedKafkaBroker;
+import org.springframework.kafka.test.context.EmbeddedKafka;
+import org.springframework.kafka.test.utils.KafkaTestUtils;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.junit.jupiter.SpringJUnitConfig;
+
+/**
+ * @author Soby Chacko
+ * @since 3.1.2
+ */
+@SpringJUnitConfig
+@DirtiesContext
+@EmbeddedKafka(topics = "enforce-rebalance-topic")
+public class ContainerEnforceRebalanceTests {
+
+	@Test
+	void enforceRebalance(@Autowired Config config, @Autowired KafkaTemplate<Integer, String> template,
+						@Autowired KafkaListenerEndpointRegistry registry) throws InterruptedException {
+		template.send("enforce-rebalance-topic", "my-data");
+		final MessageListenerContainer listenerContainer = registry.getListenerContainer("enforce-rebalance-grp");
+		assertThat(config.listenerLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(listenerContainer).isNotNull();
+		listenerContainer.enforceRebalance();
+		// The test is expecting partition revoke once and assign twice.
+		assertThat(config.partitionRevokedLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		assertThat(config.partitionAssignedLatch.await(10, TimeUnit.SECONDS)).isTrue();
+		final List<? extends KafkaMessageListenerContainer<?, ?>> containers =
+				((ConcurrentMessageListenerContainer<?, ?>) listenerContainer).getContainers();
+		assertThat(containers.get(0).enforceRebalanceRequested).isFalse();
```

Comment:
I believe `thisOrParentContainer` doesn't work for you (yet), because you check the `enforceRebalanceRequested` on the child container, but directly on the one requested from the `registry`

Reply:
I wonder why you have not fixed this part in the test...
Technically it is always `false` for any child containers in the `ConcurrentMessageListenerContainer`.
We need to be sure that it is `true` exactly in the `ConcurrentMessageListenerContainer` according to our new logic around `thisOrParentContainer`.

Or do I miss something?